### PR TITLE
added function for loading all interest categories in a list

### PR DESCRIPTION
**Pull request by bryce13950**
I need to be able to retrieve all interest categories for a project, and I thought this functionality could helper other people

### CODE REVIEW

**Comment by bryce13950** [comment]
I dont mean to be a pester or anything, but is there anyway that I can get this merged by tomorrow afternoon? I would like to wrap up a small project that I have been working on, and I need to get this into the app before I do that.

**Comment by klimov-paul** [comment]
Having ability to get "interest categories" is a good idea.

However I think such particular implementation make a little practical sense.
In my project I usually define a separated service class `InterestRepository` to hold the "interest categories" list with cross-request caching.
The main goal for me is being able to manipulate interest name instead of UID in the program.

Simple querying all "interest categories" via API is not a big deal and unlikely to worth creation of the separated method.

File: src/Newsletter.php [hunk 1]
```diff
@@ -241,6 +241,38 @@ public function updateContent(string $campaignId, string $html, array $options =
         return $response;
     }
 
+    public function getListInterestCategories(int $count = 10, int $offset = 0, string $type = null, string $listName = '')
```
Comment (klimov-paul) [inline src/Newsletter.php]: I would rather set default limit to higher value, like 100 at least.
Interest category data is tiny there is no risk of memory overflow, while usually developer wishes to fetch them all.

Comment (bryce13950) [reply src/Newsletter.php]: I just pushed up a commit with the count increased to 100

**Comment by bryce13950** [comment]
Is there any legitimate reason to not add this to the trunk? It seems like there are likely multiple people that could use an easy to use function for this, since you need to category ids before you sync a user to available categories on sign up.

**Comment by freekmurze** [comment]
Thank you for your work on this. I've decided not to merge this in, because we don't need this on our own projects and don't want to maintain it.

**Comment by bryce13950** [comment]
I think you guys misunderstand the concept of “Open source”

**Comment by freekmurze** [comment]
I agree that you probably have a different understanding about the concept than me.

**Comment by bryce13950** [comment]
You guys should put a big disclaimer on your repositories to warn people against contributing. I would have published a fork nearly a year if I knew that you guys were such dicks about people adding features to your projects.

**Comment by freekmurze** [comment]
I don't see the need of becoming rude. 

We are indeed picky of what gets merged in. We only maintain what we use ourselves. Investing time in maintaining code that we don't need, doesn't make any sense to us.

With over 7000 PRs merged I think we are open to accepting contributions for things that we want to maintain.

Feel free to fork this repo, and use our work to build the solution that you need.

**Comment by freekmurze** [comment]
Thank you for your response and being sincere. From my side I am sincere as well.

I do not agree with this statement: "Open source projects are meant to grow the ease of use for everyone, not just you". My team and I spend our paid (and sometimes free time) creating open source. This software is meant to be used on our projects. We open sourced our code because we might think it is useful for others too.

But our main goal is not "to grow the ease of use for everyone". Our goal is to build quality software for ourselves. If others can use it, that's nice. If not, we don't really mind. 

When you have created your PR, you took some time working it out. For this I am grateful. For you, this is only a one time effort. When code gets merged in, we are responsible for maintaining it. This is a task that will probably last for the lifetime of the package. If we were to merge PRs that we don't need, we would probably spend more time than needed to maintain it. This would lead to us being demotivated and potentially abandoning our package altogether. This helps no one.

When a PR doesn't get merged, your work is not lost. You can fork the package and maintain the code you need yourself. You probably are more motivated than us for maintaining it, because you need it. To me it sounds like a perfect solution. On your fork, you can even add more functionalities you need, and you don't even have to wait before things get merged. That's a big win!

Personally, but of course you don't need to agree with me here, I think a great stance towards open source is one of no expectations. Be grateful for what you get, but don't get mad or disappointment if an open source project doesn't grow your way. Closing a PR or not responding to a PR is not an offensive action. Maintainers of open source don't owe you anything.  They are perfectly in their right should they choose not to merge something in, or even to ignore you. 

I hope you can appreciate me being sincere with you. I can also highly recommend reading this blogpost on entitlement: https://tommcfarlin.com/open-source-entitlements-users

**Comment by bryce13950** [comment]
I think you miss the point. You guys provide these projects which is great, but in the long run you will end up fracturing the community with numerous off shoot projects that all do different features that someone might need when we could all simply have better libraries. You guys clearly shoot down 100's of features from tons of developers. It's not just my time that you are wasting, it is everyones. It's everyone who submits a new feature to you, and it's your time with rejecting pull requests. You guys should really make it clear that submitting new features to your projects is a waste of time, and that someone should instead purse making a branch. By not doing that, you are wasting everyones time, and if you cannot manage a naturally growing project then maybe you should consider establishing an open source organization that can manage these projects. By not doing that you are both holding back the advancement of the underlying source, and also causing the whole community to be fractured. It's great that you guys provide these libraries, but no one likes a dictator.

**Comment by Gummibeer** [comment]
@bryce13950 this was an initiative PR without any upfront issue or discussion about this feature. So it's like pushing code in your customers app he never asked for and tell them to pay for it and also pay for future maintenance of the code even if they never wanted it, needed it and will never do.
If you aren't sure that a feature gets merged I can only recommend you to open an issue upfront. The issue can contain pseudo code or even real code, describe your issue you want to solve and so on. This way the maintainer of the project can tell you if they would merge such feature or provide you with a more general feature that would allow you to add your solution more easily or guide you how to do it in a different way which would still bee compatible with the package if they won't merge anything in. You can also collect feedback from the community in this issue.

And forking isn't the only solution, I know Freek and several other maintainers add links to valuable packages that work great in combination with this one. This would solve all issues all of you two have.
* Spatie doesn't have to maintain code they never use
* You get your code as soon as you've written it
* You don't have to maintain the fork
* Users don't have to decide which fork they use as they add "plugins" to the original package

**Comment by bryce13950** [comment]
The amount of code that spatie provides to the community is absolutely incredible. They are without question one of the largest laravel package contributors around, but I think it's pretty obvious that this has lead to gate keeping. I asked in June if there was any reason to not merge this. It took over three months to get a response. If at any point in the last year someone from spatie had informed me that they did not have enough resources to maintain this then I would have personally volunteered to help maintain this package, and many more spatie packages for the betterment of the community. Having this being rejected due to a lack of willingness to maintain it makes me wonder how many other features have ended up as forks in the 300+ libraries spatie has created. It's one thing to reject a pull request because of a lack of following standards, a lack of testing, or from bad organization, but this is just silly.

Continuing to use my fork in production is fine. It's just a really dumb thing to have to do when the little bitty function I added could help other people without adding anything negative to the overall project. All of these libraries are probably splintered multiple times like this. If spatie is unable to maintain these packages in an open way, then I really believe that they should establish a community ran open source organization to maintain it. A good example of this would be the ionic or WordPress projects. As is, the lack of flexibility in these libraries have probably evolved to a point where it is hurting the future evolution of the community.

Also, everything said in the last comment @Gummibeer should be published in this repo. It seems like there is a pretty clear expectation on how to accept changes in the project, but that is not written anywhere in the actual source base of the project.

**Comment by Gummibeer** [comment]
@bryce13950 I can't speak for Freek or any other spatie member but in general they are adding "external" maintainers for packages - as I'm for ~5 of them and a massive contributor in a lot more.

In general I do this in the spatie and my own packages that features without an issue are hard to accept as the issue to solve is missing. (not talking about simple things)
I don't know what, where or how you work but in our company we have team meeting to plan the open tasks, in these meetings we talk about some namings, if/which packages to use/create and so on. During these talks we get the first missing details in the expectation like incomplete validation and are all on a similar common-sense of how the final solution should look like.
Open-Source needs these discussions too, but in opposite to a team sitting together in a single room they have to be done via the GitHub issues (or whichever platform the maintainer chooses). This will allow the maintainer to think about it before any code is written, it also allows the maintainer to check with other open issues and features on the roadmap how they will fit together.
For example making the newsletter macroable in this package would nearly introduce no future maintenance overhead but make it easier for others to add functionality to the class. Possibly this could have been a solution - possibly it still is a solution. 😉
Sometimes open-source means to reject a detailed feature in favor of a more general implementation which allows the user to add the requested feature and a lot more. We did something similar in the activitylog package by adding the `tap()` logic.
https://spatie.be/docs/laravel-activitylog/v3/advanced-usage/logging-model-events#tap-activity-before-logged-from-event
It was one of the best features (from the maintainers perspective) we've added as it allowed us to close nearly every issue with the tap logic as it's extremely powerful. PS: we've canceled/rejected ~10-20 features requests with this.
So sometimes going for a more general solution is a better solution for all as it prevents adding 10 more detailed features.

I agree that response times are very different across spatie packages. As a little explanation/excuse: being the owner/maintainer of this amount of packages means around `rand(1, 10) * packages.count()` amount of GitHub notifications per day. So it's nearly impossible to keep track of all of them. On top of the "important" notifications from your paid projects.
https://twitter.com/freekmurze/status/1303776566439219200
I could imagine that Spatie isn't publicly searching for maintainers as it would only lead to a hundred more notifications per day from all the devs wanting to b able to tell everyone that they work for/with spatie - like as you said it's a known company/name in Laravel universe.

And I totally know how disappointing it can be if your work/code/PR is stuck somewhere on GitHub. But Open-Source maintainers are also only humans who have to get work done, have holidays, get ill and sometimes simply have a bad day and don't want to care. And yes, sometimes it hits the "wrong" - as after you've answered on 10 silly issues with "read the *** docs" (deleted everything and answered with a more kind wording 😉) you open the next link and just close it because you are done for the day.
This isn't a real justification, possibly also because there isn't a real need for one. As Freek said no one owes anything to anyone in Open-Source. But possibly a chance to get an idea of what it means to maintain this amount of code.

If you have any ideas, except for more humans on it, I'm open to hear them! 🙂
And if you have any ideas on how to get more humans on it, without wasting a ton of time with spam applications, I'm also open! And if you are really willing to maintain one or more packages - head over to a direct chat with Freek.

**Comment by bryce13950** [comment]
@Gummibeer All of your suggestions are great. There are a lot of ways that communication could be improved without simply adding more people. As we have discussed, the onboard process that you have laid out in your comments is something very clear and easy to follow, but it is completely nonexistent in any of the docs within the repo. Before starting to work on this little feature I reviewed all docs in the repo, and I found nothing on the manner in which to contribute a feature. Simply adding a contribution guide to the repos could easily help organize the workload, and funnel requests through a process that is more manageable. 

Also, if the maintainer is getting that many notifications a day, then maybe a big problem is that things aren't being delegated enough. That is a much deeper problem though, but I am sure that there are reorganization projects that could properly be executed to spread that load.

**Comment by Gummibeer** [comment]
I agree that this is missing - #Hacktoberfest get it fixed. 😉 (as I'm not a maintainer of this project that's only my opinion and not real advice that will guarantee a merge^^)
https://github.com/spatie/laravel-newsletter/blob/master/.github/CONTRIBUTING.md
But something like "for larger features we recommend to discuss the issue to solve first before committing code that we will probably reject" should be fine @freekmurze !? 🤔
I'm also not sure if @rubenvanassche  or anyone else at spatie has probably a script in place to spread this change over all packages!?
PS: this could also be solved by a central `spatie/.github` repository - we do something similar https://github.com/Astrotomic/.github

**Pull request closed without merging**